functions/TokenGenerator/functions/index.js

```diff
@@ -0,0 +1,179 @@
+      },
+    };
+
+    // Set up the request
+    let oauthToken = '';
+    const post_req = https.request(post_options, res => {
+      res.setEncoding('utf8');
+      res.on('data', chunk => {
+        oauthToken += chunk;
+      });
+      res.on('end', () => {
+        // Next step in pipeline
+        saveOAuthToken(JSON.parse(oauthToken));
+        return resolve(JSON.parse(oauthToken));
+      });
+    });
+
+    post_req.on('error', e => {
+      console.log('ERROR generating new token', e.message);
+      return 'Error retrieving token';
+    });
+
+    // Sets up the scope that we want the end user to have.
+    const body = {
+      delegates: [],
+      scope: ['https://www.googleapis.com/auth/' + scope],
+      lifetime: '3599s',
+    };
+
+    // post the data
+    post_req.write(JSON.stringify(body));
+    post_req.end();
+  });
+}
+// [END generate_token]
+
+// [START retrieve_credentials]
+function retrieveCredentials(projectID) {
+  return new Promise(resolve => {
+    // To create a new access token, we first have to retrieve the credentials
+    // of the service account that will make the generateTokenRequest().
+    // To do that, we will use the App Engine Default Service Account.
+    const options = {
+      host: 'metadata.google.internal',
+      path: '/computeMetadata/v1/instance/service-accounts/default/token',
+      method: 'GET',
+      headers: {'Metadata-Flavor': 'Google'},
+    };
+
+    const get_req = http.get(options, res => {
+      let body = '';
+
+      res.on('data', chunk => {
+        body += chunk;
+      });
+
+      res.on('end', () => {
+        const response = JSON.parse(body);
+        return generateAccessToken(
+          projectID,
+          response.access_token,
+          response.token_type
+        ).then(result => {
+          return resolve(result);
+        });
+      });
+    });
+    get_req.on('error', e => {
+      //console.log('Error retrieving credentials', e.message);
+      return 'Error retrieving token' + e.message;
+    });
+    get_req.end();
+  });
+}
+exports.retrieveCredentials = retrieveCredentials;
+// [END retrieve_credentials]
+
+// [START validate_token]
+// This method verifies the token expiry by validating against current time
+function isValid(expiryTime) {
+  const currentDate = new Date();
+  const expirationDate = new Date(expiryTime);
+  // If within 5 minutes of expiration, return false
+  return currentDate <= expirationDate - 1000 * 60 * 5;
+}
+// [END validate_token]
+
+// [START function_get_token]
+exports.getOAuthToken = functions.https.onCall((data, context) => {
+  // Checking that the user is authenticated.
+  if (!context.auth) {
+    // Throwing an HttpsError so that the client gets the error details.
+    throw new functions.https.HttpsError(
+      'failed-precondition',
+      'The function must be called ' + 'while authenticated.'
+    );
+  }
+  const projectID = data["projectID"];
```

Comment:
I think this might be fine, but should consider the implications if a user can change this variable and what abuse it opens the function to.